chronos/achannels.nim

```diff
@@ -0,0 +1,254 @@
+#
+#    Chronos multi-threading asynchronous channels
+#             (c) Copyright 2018-Present
+#         Status Research & Development GmbH
+#
+#              Licensed under either of
+#  Apache License, version 2.0, (LICENSE-APACHEv2)
+#              MIT license (LICENSE-MIT)
+import strutils
+import asyncloop, handles, transport, asyncsync
+
+const hasThreadSupport* = compileOption("threads")
+
+when hasThreadSupport:
+  import locks
+
+type
+  RawAsyncChannelImpl {.pure, final.} = object
+    rd, wr, count, mask: int
+    maxItems: int
+    refCount: int
+    data: ptr UncheckedArray[byte]
+    when hasThreadSupport:
+      lock: Lock
+    eventNotEmpty: AsyncThreadEvent
+    eventNotFull: AsyncThreadEvent
+
+  RawAsyncChannel = ptr RawAsyncChannelImpl
+  AsyncChannel*[Msg] = RawAsyncChannel
+  AsyncChannelError* = object of CatchableError
+
+proc initLocks(rchan: RawAsyncChannel) {.inline.} =
+  ## Initialize and create OS locks.
+  when hasThreadSupport:
+    initLock(rchan.lock)
+  else:
+    discard
+  rchan.eventNotEmpty = newAsyncThreadEvent()
+  if rchan.maxItems > 0:
+    rchan.eventNotFull = newAsyncThreadEvent()
+
+proc deinitLocks(rchan: RawAsyncChannel) {.inline.} =
+  ## Deinitialize and close OS locks.
+  when hasThreadSupport:
+    deinitLock(rchan.lock)
+  else:
+    discard
+
+  close(rchan.eventNotEmpty)
+  if rchan.maxItems > 0:
```

Comment:
Why is this conditional?

Reply:
Because if `rchan.maxItems <= 0`, then queue is become infinite and second event `notfull` will be not created and used, so we going to close it only for limited queue.